### ISSUE TITLE
[Feature] run tests only against changed files

### ISSUE THREAD

**Issue by SimonGolms** [microsoft/playwright#15075]
## Description

As a user, I want to run my component tests only on changed files, as it would be a use case in a git hook ([`pre-push`](https://git-scm.com/docs/githooks#_pre_push))

## Alternatives

- https://jestjs.io/docs/cli#--onlychanged
- https://vitest.dev/guide/cli.html#changed

## Preferred Solution

Enhancement of the cli commands, e.g.

```sh
npx playwright test --only-changed
```

**Comment by jfgreffier** [comment]
I propose something inspired by Vitest, that can check for uncommited changes, or changes compared to a commit.

```bash
npx playwright test --changed
npx playwright test --changed d819f97f40386c9bd31a53525240c6e9997a10f4
npx playwright test --changed main
```

What do you think ?

**Comment by jfgreffier** [comment]
WIP, I just pushed documentation

**Comment by jfgreffier** [comment]
The PR is completed since a while. Waiting for feedback and review.

**Comment by joeflan** [comment]
> The PR is completed since a while. Waiting for feedback and review.

I saw the PR was closed. Any chance of re-opening and taking a look at this again? It's something that would greatly benefit our team

**Comment by y-nk** [comment]
what's the status here? opened but not merged. can someone have a look? it seems like a pretty nice feature.

**Comment by sayers24** [comment]
@pavelfeldman - Looks like you had feedback before closing the PR. Is there any direction? My team is blocked on merging tests due to untouched tests failing.

**Comment by max-mykhailenko** [comment]
Waiting for this so much. Do you have it in your roadmap?

**Comment by visualjeff** [comment]
This would be helpful for our organization.   We don't need to run tests on stories that haven't changed.  Any updates on this feature?

**Comment by Proryanator** [comment]
Also would like this too, having moved from Selenium to Playwright, this would be a huge feature. It is easy to fetch changed `*.spec.ts` files with git via some cli stuff, but also having it just built in, with support for also identifying used objects to determine what tests to run would be amazing.

**Comment by MindaugasMateika** [comment]
> Also would like this too, having moved from Selenium to Playwright, this would be a huge feature. It is easy to fetch changed `*.spec.ts` files with git via some cli stuff, but also having it just built in, with support for also identifying used objects to determine what tests to run would be amazing.

Problem is this feature requires two options / implementations which is more complex than just parsing git diff. One option to run changed tests based on test code change, another is to run tests based on, for example, component code changes like what Jest is doing with `--only-changed` option.

**Comment by gkiely** [comment]
Relevant code: https://github.com/jestjs/jest/tree/main/packages/jest-changed-files

**Comment by Skn0tt** [comment]
#31727 ships a version of this that doesn't rely on any external libraries. It'll be included in the next minor release.

It doesn't yet support UI mode. If that's something you're interested in, please open a new issue referencing this one. Thanks!

(this is the first feature i've worked on since joining the Playwright team last week, so it'll have a place in my heart forever!)

**Comment by ZooDoo4U** [comment]
Might make a suggestion.  Including the PW version for which the "Last Change" considers.   Example a test run once and never changes.  Possibly the test will never run again as it passed the last time and never changed.  Just asking that at a minimum a change in the PW will force the test to re-run at least once per version...?

**Comment by Skn0tt** [comment]
Hey Dan! I like the idea, but I think it adds undue complexity. Let me explain how we think about the feature for some context:

`--only-changed` is a heuristic, and it'll never be able to 100% understand what code changes need what tests to rerun. Think of application changes that don't manifest as code changes, like when an upstream API changes its schema.
Because of that, `--only-changed` will never be able to 100% understand what tests need to be rerun.
To continue testing safely, this means that you'll need to run your entire Suite *at some point*.

We think of `--only-changed` as being useful in three distinct situations:
1. **Local Development:** it's similar to `--watch`
2. **Saving *time* in CI:** running tests with `--only-changed` *before* doing a full run leads to faster failures on modified tests
3. **Saving *money* in CI:** running with `--only-changed` on PRs and doing a full run every night. Leads to less test runs at the expense of a longer feedback loop

In either of those cases, the full suite is executed at some point, and that will catch any changes in the Playwright version. So I don't think we'll need to take it into account.

All of the above will go into the docs btw, working on that right now!

**Comment by ZooDoo4U** [comment]
@Skn0tt Thanks for the comments, just wanted to make sure there wasn't an unexpected test hole was created.   Guess no job is finished til the paperwork is done 🤪
